flink-clients/src/main/java/org/apache/flink/client/cli/ExecutorCLI.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Properties;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A generic implementation of the {@link CustomCommandLine} that only expects
+ * the execution.target parameter to be explicitly specified and simply forwards the
+ * rest of the options specified with -D to the corresponding {@link org.apache.flink.core.execution.Executor}
+ * for further parsing.
+ */
+@Internal
+public class ExecutorCLI implements CustomCommandLine {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ExecutorCLI.class);
+
+	private static final String ID = "Executor-CLI";
+
+	private final Option executorOption = new Option("e", "executor", true,
+			"The name of the executor to be used for executing the given job, e.g. \"local-executor\"." +
+					" This is equivalent to the \"" + DeploymentOptions.TARGET.key() + "\" config option.");
+
+	/**
+	 * Dynamic properties allow the user to specify additional configuration values with -D, such as
+	 * <tt> -Dfs.overwrite-files=true  -Dtaskmanager.memory.shuffle.min=536346624</tt>.
+	 */
+	private final Option dynamicProperties = Option.builder("D")
+			.argName("property=value")
+			.numberOfArgs(2)
+			.valueSeparator('=')
+			.desc("use value for given property")
+			.build();
+
+	private final Configuration baseConfiguration;
+
+	public ExecutorCLI(final Configuration configuration) {
+		this.baseConfiguration = new UnmodifiableConfiguration(checkNotNull(configuration));
+	}
+
+	@Override
+	public boolean isActive(CommandLine commandLine) {
+		return baseConfiguration.getOptional(DeploymentOptions.TARGET).isPresent()
+				|| commandLine.hasOption(executorOption.getOpt());
+	}
+
+	@Override
+	public String getId() {
+		return ID;
+	}
+
+	@Override
+	public void addRunOptions(Options baseOptions) {
+		// nothing to add here
+	}
+
+	@Override
+	public void addGeneralOptions(Options baseOptions) {
+		baseOptions.addOption(executorOption);
+		baseOptions.addOption(dynamicProperties);
+	}
+
+	@Override
+	public Configuration applyCommandLineOptionsToConfiguration(final CommandLine commandLine) {
+		checkState(isActive(commandLine));
+
+		final Configuration effectiveConfiguration = new Configuration(baseConfiguration);
+
+		final String executorName = commandLine.getOptionValue(executorOption.getOpt());
+		if (executorName != null) {
+			effectiveConfiguration.setString(DeploymentOptions.TARGET, executorName);
+		}
+
+		encodeDynamicProperties(commandLine, effectiveConfiguration);
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("Effective Configuration: {}", effectiveConfiguration);
+		}
+
+		return effectiveConfiguration;
+	}
+
+	private void encodeDynamicProperties(final CommandLine commandLine, final Configuration effectiveConfiguration) {
```

Comment:
We actually change the behavior here that, for example, previously a user can configure image via `-i` or `--image` option but now it can be only configured as `-Dkubernetes.container.image=...`.

CC @wangyang0918

Reply:
Yes, you are right. We already have some discussion under https://issues.apache.org/jira/browse/FLINK-15179. Looking forward to your opinion.

Reply:
Aha. Given the PR is present I don't checkout the JIRA. Will take a look later.